src/discussions/speed-grader.ts

```diff
@@ -0,0 +1,178 @@
+        }
+
+        this.observer = new MutationObserver(() => this.handleIframeHolderMutations());
+        this.observer.observe(iframeHolder, { attributes: true, childList: true, subtree: false });
+    }
+
+    private setStudentId(): void {
+        const currentLocation = window.location.href;
+        const regex = /speed_grader.*student_id=(\d+)/;
+        const match = currentLocation.match(regex);
+
+        if (match) {
+            const [, studentId] = match;
+            this.studentId = studentId;
+        }
+    }
+
+    private handleIframeHolderMutations(): void {
+        const iframe = document.getElementById('speedgrader_iframe') as HTMLIFrameElement;
+        if (!iframe) {
+            this.log.info('Speedgrader iframe not found');
+            return;
+        }
+
+        const dom = iframe.contentDocument;
+        if (!dom) {
+            this.log.info('Cannot access iframe content document');
+            return;
+        }
+
+        let attempt = 0;
+        const userContentLoadInterval = setInterval(() => {
+            const userContent = iframe.contentDocument?.getElementsByClassName('user_content');
+            if (userContent?.length) {
+                this.userContentLoadedHandler(iframe);
+                clearInterval(userContentLoadInterval);
+            } else {
+                attempt += 1;
+                if (attempt >= MAX_ATTEMPTS) {
+                    this.log.info('Failed to load user content after maximum attempts');
+                    clearInterval(userContentLoadInterval);
+                }
+            }
+        }, USER_CONTENT_LOADED_INTERVAL);
+    }
+
+    private userContentLoadedHandler(speedgraderIframe: HTMLIFrameElement): void {
+        const dom = speedgraderIframe.contentDocument;
+        if (!dom) {
+            this.log.info('Cannot access iframe content document');
+            return;
+        }
+        const discussionLinkElement = dom.getElementById(
+            'discussion_view_link'
+        ) as HTMLAnchorElement;
+        if (!discussionLinkElement) {
+            this.log.info('Discussion link not found');
+            return;
+        }
+
+        const discussionTopicLink = discussionLinkElement.href;
+        const regex = /courses\/(\d+)\/discussion_topics\/(\d+)/;
+        const matches = discussionTopicLink.match(regex);
+        if (!matches) {
+            this.log.info('Discussion topic details not found');
+            return;
+        }
+
+        const courseNumber = matches[1];
+        const topicNumber = matches[2];
+
+        const iframes = dom.querySelectorAll('iframe');
+        iframes.forEach((iframe, key) =>
+            this.iframeHandler(iframe as HTMLIFrameElement, key, courseNumber, topicNumber)
+        );
+    }
+
+    private iframeHandler(
+        iframe: HTMLIFrameElement,
+        key: number,
+        courseNumber: string,
+        topicNumber: string
+    ): void {
+        if (!iframe.src.includes('external_tools')) {
```

Comment:
what if it's not Annoto?

Reply:
why not implement filter at iframes.forEach().filter

Reply:
It is iterator, but you are right, it will be more reasonable to filter before call

Reply:
Unfortunately I didn't find any option to understand that it is Annoto:
![image](https://github.com/user-attachments/assets/f87e5f8c-9ab5-4f7c-8312-0e38d57d645a)
It was a possible solution to limit attempt amount to iframe, but canvas doesn't load internals of iframe until scroll and I couldn't understand is it another external tool, or it just not loaded
Example with not loaded annoto iframe
![image](https://github.com/user-attachments/assets/e507f0b9-4c02-40e3-a685-01c5e961f222)